crates/e2e/src/backend_calls.rs

```diff
@@ -0,0 +1,366 @@
+use ink_env::Environment;
+use pallet_contracts_primitives::ContractInstantiateResult;
+use scale::{
+    Decode,
+    Encode,
+};
+use sp_weights::Weight;
+
+use crate::{
+    backend::BuilderClient,
+    builders::CreateBuilderPartial,
+    CallBuilderFinal,
+    CallDryRunResult,
+    CallResult,
+    ContractsBackend,
+    InstantiationResult,
+    UploadResult,
+};
+
+use super::Keypair;
+
+/// Allows to build an end-to-end call using a builder pattern.
+pub struct CallBuilder<'a, E, Args, RetType, B>
+where
+    E: Environment,
+    Args: Encode + Clone,
+    RetType: Send + Decode,
+
+    B: BuilderClient<E>,
+{
+    client: &'a mut B,
+    caller: &'a Keypair,
+    message: &'a CallBuilderFinal<E, Args, RetType>,
+    value: E::Balance,
+    extra_gas_portion: Option<u64>,
+    gas_limit: Option<Weight>,
+    storage_deposit_limit: Option<E::Balance>,
+}
+
+impl<'a, E, Args, RetType, B> CallBuilder<'a, E, Args, RetType, B>
+where
+    E: Environment,
+    Args: Sync + Encode + Clone,
+    RetType: Send + Decode,
+
+    B: BuilderClient<E>,
+{
+    /// Initialize a call builder with defaults values.
+    pub fn new(
+        client: &'a mut B,
+        caller: &'a Keypair,
+        message: &'a CallBuilderFinal<E, Args, RetType>,
+    ) -> CallBuilder<'a, E, Args, RetType, B>
+    where
+        E::Balance: From<u32>,
+    {
+        Self {
+            client,
+            caller,
+            message,
+            value: 0u32.into(),
+            extra_gas_portion: None,
+            gas_limit: None,
+            storage_deposit_limit: None,
+        }
+    }
+
+    /// Provide value with a call
+    pub fn value(&mut self, value: E::Balance) -> &mut Self {
```

Comment:
why not `pub fn value(mut self, value: E::Balance) -> Self`?

Reply:
We don't want to move values out of references, and, generally speaking, it is a good practise to implement non-consuming builders where possible:
https://doc.rust-lang.org/1.0.0/style/ownership/builders.html#non-consuming-builders-(preferred):